nixos/qemu-vm: add option for named network interfaces

GrahamDennis opened the pull request:
Adds a new option to the virtualisation modules that enables specifying explicitly named network interfaces in QEMU VMs. The existing virtualisation.vlans is still supported for cases where the name of the network interface is irrelevant.

This feature is useful for creating VMs with network configurations that exactly match the network configuration of real-world devices for more accurate SITL testing.

Note: This is PR https://github.com/NixOS/nixpkgs/pull/178290 which was later reverted in https://github.com/NixOS/nixpkgs/pull/213519 because it was causing some hydra tests to fail: https://github.com/NixOS/nixpkgs/pull/178290#issuecomment-1407751385.

This PR includes a fix for the "sit" and "privacy" tests that failed after the PR was merged last time, and adds a new test to cover the situation where the MAC addresses in the QEMU VM contain a character in the range [a-f].

###### Description of changes

* Added virtualisation.interfaces option to qemu-vm.nix.
* Updated the default value of virtualisation.vlans to empty if virtualisation.interfaces is not empty. This avoids a scenario where a user specifies virtualisation.interfaces, but inadvertently still has a network interface from the default virtualisation.vlans.
* Updated build-vms.nix to create the new network interfaces using QEMU options.
* Updated build-vms.nix to rename network interfaces using udev rules.
* Updated testing-python.nix to use the union of the VLANs specified in virtualisation.vlans and virtualisation.interfaces when collecting the list of VLANs.
* Updated networking tests in nixos/tests/networking.nix to use the new option where applicable. This cut down on unnecessary overrides to remove IP addresses from network interfaces.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
nixos/qemu-vm: add option for named network interfaces

Adds a new option to the virtualisation modules that enables specifying explicitly named network interfaces in QEMU VMs.
The existing `virtualisation.vlans` option is still supported for cases where the name of the network interface is irrelevant.

RaitoBezarius commented:
Awesome, please tag me for such PRs if you don't want specialized reviewers to miss them. :)

GrahamDennis commented:
Thanks @RaitoBezarius! I have rebased and squashed my commits amending the commit title (and PR title) to match the guidelines.

I'll also need to move my change to the release notes after 23.05 branches.

RaitoBezarius commented:
> Am I right in understanding that hydra hasn't run the PR yet? It looks to have unrelated evaluation errors.

At the moment, I see `ofborg-eval` cloning the project, so not running it yet.

tfc commented on nixos/lib/testing/network.nix:
```diff
@@ -18,24 +18,41 @@ let
+        qemu-common.qemuNICFlags snd fst.vlan config.virtualisation.test.nodeNumber));
+      udevRules = forEach interfacesNumbered ({ fst, snd }:
+        # MAC Addresses for QEMU network devices are lowercase, and udev string comparison is case-sensitive.
+        "SUBSYSTEM==\"net\",ACTION==\"add\",ATTR{address}==\"${toLower(qemu-common.qemuNicMac fst.vlan config.virtualisation.test.nodeNumber)}\",NAME=\"${fst.name}\"");
```
This is just a nitpick, feel free to ignore, but if using `'''strings'''` we would have less backslash-noise in this rule.

GrahamDennis replied:
done. I've also rebased on master after the 23.05 release branched.

Commit:
nixos/qemu-vm: quoted string reformat

GrahamDennis commented:
I don't understand why the manual check is failing. Does anyone have ideas?

RaitoBezarius commented:
Potentially because I fucked up the initialization of the manual :-)

RaitoBezarius commented:
If you can solve the conflicts, it'd be awesome. :)

GrahamDennis commented:
Done @RaitoBezarius!

RaitoBezarius commented:
> Done @RaitoBezarius!

I'm evaluating a recent nixos-unstable-small and we will evaluate this PR over https://hydra.nixos.org/jobset/nixos/python-test-refactoring to smoke out any regression. :)

RaitoBezarius commented:
You should see one of your commit here: https://hydra.nixos.org/jobset/nixos/python-test-refactoring in ~30 mn. It will take some amount of time as the queue is currently full of other jobs and we don't have any priority.

Let's rediscuss when we will have some new (non flaky) failures.

RaitoBezarius commented:
Please review the 4 new failures here: https://hydra.nixos.org/eval/1795393

I think aarch64 are classical spurious failures, please double check the x86_64 one.

If we are good, I think this is good to go.

GrahamDennis commented:
None of those failures seem related to this change. The x86_64 change is a timeout, and it looks like control never passed from the kernel to userspace (https://hydra.nixos.org/build/221399319/log). I think we are good.
I wasn't aware that NixPkgs has flaky tests... this must be very frustrating

RaitoBezarius reviewed (commented):
(was already reviewed in the previous PR)
